source/material_model/rheology/viscosity_prefactors.cc

```diff
@@ -0,0 +1,126 @@
+/*
+  Copyright (C) 2020 - 2023 by the authors of the ASPECT code.
+
+  This file is part of ASPECT.
+
+  ASPECT is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation; either version 2, or (at your option)
+  any later version.
+
+  ASPECT is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with ASPECT; see the file LICENSE.  If not see
+  <http://www.gnu.org/licenses/>.
+*/
+
+
+#include <aspect/material_model/rheology/viscosity_prefactors.h>
+#include <aspect/utilities.h>
+#include <aspect/global.h>
+#include <aspect/geometry_model/interface.h>
+
+#include <deal.II/base/signaling_nan.h>
+#include <deal.II/base/parameter_handler.h>
+#include <aspect/simulator_signals.h>
+#include <aspect/parameters.h>
+
+
+// TODO: Add constant_viscosity_prefactors to this file at a point when ASPECT is ready to be broken
+namespace aspect
+{
+  namespace MaterialModel
+  {
+    namespace Rheology
+    {
+      template <int dim>
+      ViscosityPrefactors<dim>::ViscosityPrefactors ()
+        = default;
+
+
+
+      template <int dim>
+      double
+      ViscosityPrefactors<dim>::compute_viscosity (const double base_viscosity,
+                                                   const unsigned int composition_index) const
+      {
+        double time = this->get_time();
+        if (this->convert_output_to_years())
+          time /= year_in_seconds;
+
+        const double factor = weakening_times[composition_index] <= time ? variable_viscosity_prefactors[composition_index] : 1.0;
```

Comment:
I'm a bit confused by this new functionality. Our discussion earlier focused on state-dependent viscosity (i.e., viscosities varying as a function of composition in ways not covered by existing material averaging). 

As written, the new function weakens as a function of time globally within the domain. Do you have a use-case in mind for this?